C++ API parity: isfinite

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28783 C++ API parity: BCEWithLogitsLoss
* **#28918 C++ API parity: isfinite**

Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

pbelevich committed:
C++ API parity: isfinite

[ghstack-poisoned]

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -1603,3 +1603,29 @@ TEST_F(FunctionalTest, PoissonNLLLoss) {
+    const auto inf = std::numeric_limits<T>::infinity();
+    const auto x = torch::full({3, 3}, inf, torch::TensorOptions().dtype(S));
+    ASSERT_FALSE(torch::isfinite(x).all().template item<bool>());
+  }
```
Would be awesome to add a test where only some of the tensor elements are finite :D

yf225 commented on torch/csrc/api/include/torch/functional.h:
```diff
@@ -0,0 +1,14 @@
+  if (!tensor.is_floating_point()) {
+    return torch::ones_like(tensor, torch::kBool);
+  }
+  return (tensor == tensor) * (tensor.abs() != torch::full_like(tensor, std::numeric_limits<double>::infinity()));
```
When the `tensor` is of `float` dtype, would `tensor.abs() != torch::full_like(tensor, std::numeric_limits<double>::infinity())` always evaluate to true even if `tensor` is full of `std::numeric_limits<float>::infinity()` value? Or should we dispatch the `tensor.abs() != torch::full_like(tensor, std::numeric_limits<double>::infinity())` check based on tensor dtype (i.e. using macros like `AT_DISPATCH_ALL_TYPES`)?

zou3519 commented on torch/csrc/api/include/torch/functional.h:
```diff
@@ -0,0 +1,14 @@
+
+namespace torch {
+
+inline at::Tensor isfinite(const at::Tensor& tensor) {
```
We should move the implementation of python isfinite to c++ unless there is some reason why the python version cannot call into c++. This way, we'll have just one implementation for both: https://pytorch.org/docs/master/_modules/torch/functional.html#isfinite.

yf225 replied:
Yes agreed that it would be awesome to unify these two.

zou3519 replied:
To clarify, to do this, one would:
- add a new `isfinite` entry into native_functions.yaml: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml
- delete the `torch/functional.py` version of `isfinite` (it won't be  necessary anymore)
- while preserving all documentation (the torch/functional.py isfinite docs might need to be moved to _torch_docs.py.

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -87,6 +87,13 @@ Tensor isnan(const Tensor& self) {
 }
 
+Tensor isfinite(const Tensor& self) {
+  if (!self.is_floating_point()) {
```
Might be good to include part of the original comment, i.e., "Integral tensor types are finite"

zou3519 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -87,6 +87,13 @@ Tensor isnan(const Tensor& self) {
+  if (!self.is_floating_point()) {
+    return at::ones_like(self, at::kBool);
+  }
+  return (self == self) * (self.abs() != at::full_like(self, std::numeric_limits<double>::infinity()));
```
It's faster to just compare with `at::full(1, std::numeric_limits<double>::infinity()))`
Also, is double infinity the same thing as float infinity?

pbelevich replied:
done

zou3519 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -87,6 +87,13 @@ Tensor isnan(const Tensor& self) {
 
+Tensor isfinite(const Tensor& self) {
+  if (!self.is_floating_point()) {
+    return at::ones_like(self, at::kBool);
```
Does it matter that we're not using `memory_format=torch.legacy_contiguous_format`?

pbelevich replied:
changed to `at::MemoryFormat::Preserve` after discussing with Vitaly

zou3519 commented on torch/csrc/api/include/torch/functional.h:
```diff
@@ -0,0 +1,11 @@
+
+namespace torch {
+
+inline at::Tensor isfinite(const at::Tensor& tensor) {
```
Do we still need this file? I thought that all aten functions get bound into the torch namespace, so `torch::isfinite` might exist already

pbelevich replied:
deleted

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -87,6 +87,16 @@ Tensor isnan(const Tensor& self) {
+    return at::ones_like(self, at::kBool, at::MemoryFormat::Preserve);
+  }
+  return AT_DISPATCH_FLOATING_TYPES_AND_HALF(self.scalar_type(), "isfinite", [&]() {
+    return (self == self) * (self.abs() != at::full(1, std::numeric_limits<scalar_t>::infinity(), self.options()));
```
Does `self.abs() != std::numeric_limits<double>::infinity()` work? I'm not sure how the `Scalar` API works or if matters at all.

pbelevich replied:
Looks like it works

zou3519 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -87,6 +87,16 @@ Tensor isnan(const Tensor& self) {
+Tensor isfinite(const Tensor& self) {
+  // Integral tensor types are finite
+  if (!self.is_floating_point()) {
+    return at::ones_like(self, at::kBool, at::MemoryFormat::Preserve);
```
This used to be `torch.legacy_contiguous_format`; is this change OK?

pbelevich replied:
I discussed it with Vitaly and he suggested to make it `at::MemoryFormat::Preserve`

zou3519 commented on test/cpp/api/functional.cpp:
```diff
@@ -1960,3 +1960,62 @@ TEST_F(FunctionalTest, Dropout3d) {
+      inf
+    }, torch::TensorOptions().dtype(S).device(device));
+    ASSERT_TRUE(torch::allclose(
+      torch::isfinite(x).toType(torch::kInt),
```
We're taking a BoolTensor, converting to IntTensor, and then comparing it to another IntTensor that was converted from a BoolTensor. Is there a way to just directly compare the BoolTensors ?

pbelevich replied:
torch::allclose does not support torch::kBool

zou3519 replied:
Could you toss this information into an in-line comment in the testing code?

pbelevich replied:
done

zou3519 reviewed (commented):
Some questions

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: isfinite"


Differential Revision: [D18233037](https://our.internmc.facebook.com/intern/diff/D18233037)

[ghstack-poisoned]

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -1960,3 +1960,63 @@ TEST_F(FunctionalTest, Dropout3d) {
+}
+
+TEST_F(FunctionalTest, isfinite) {
+  for (const auto device : {at::Device("cpu"), at::Device("cuda")}) {
```
CPU-only tests seem to fail with
```
Nov 19 00:33:35 [ RUN      ] FunctionalTest.isfinite
Nov 19 00:33:35 unknown file: Failure
Nov 19 00:33:35 C++ exception with description "Cannot initialize CUDA without ATen_cuda library. PyTorch splits its backend into two shared libraries: a CPU library and a CUDA library; this error has occurred because you are trying to use some CUDA functionality, but the CUDA library has not been loaded by the dynamic linker for some reason.  The CUDA library MUST be loaded, EVEN IF you don't directly use any symbols from the CUDA library! One common culprit is a lack of -Wl,--no-as-needed in your link arguments; many dynamic linkers will delete dynamic library dependencies if you don't depend on any of their symbols.  You can check if this has occurred by using ldd on your binary to see if there is a dependency on *_cuda.so library. (initCUDA at /var/lib/jenkins/workspace/aten/src/ATen/detail/CUDAHooksInterface.h:63)
Nov 19 00:33:35 frame #0: <unknown function> + 0xcfa83 (0x7fabe4f6ca83 in /var/lib/jenkins/workspace/build/lib/libc10.so)
Nov 19 00:33:35 frame #1: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x1c8 (0x7fabe4f6b118 in /var/lib/jenkins/workspace/build/lib/libc10.so)
Nov 19 00:33:35 frame #2: <unknown function> + 0xb1c616f (0x7fabf03e916f in /var/lib/jenkins/workspace/build/lib/libtorch.so)
Nov 19 00:33:35 frame #3: <unknown function> + 0xb12b6a2 (0x7fabf034e6a2 in /var/lib/jenkins/workspace/build/lib/libtorch.so)
Nov 19 00:33:35 frame #4: <unknown function> + 0xea99 (0x7fabdd57ba99 in /lib/x86_64-linux-gnu/libpthread.so.0)
Nov 19 00:33:35 frame #5: <unknown function> + 0xb12b363 (0x7fabf034e363 in /var/lib/jenkins/workspace/build/lib/libtorch.so)
Nov 19 00:33:35 frame #6: <unknown function> + 0xea99 (0x7fabdd57ba99 in /lib/x86_64-linux-gnu/libpthread.so.0)
Nov 19 00:33:35 frame #7: <unknown function> + 0xbcffbc (0x56250335cfbc in build/bin/test_api)
Nov 19 00:33:35 frame #8: <unknown function> + 0xbea6f1 (0x5625033776f1 in build/bin/test_api)
Nov 19 00:33:35 frame #9: <unknown function> + 0xbb270e (0x56250333f70e in build/bin/test_api)
Nov 19 00:33:35 frame #10: void test_isfinite<(c10::ScalarType)0, unsigned char>(c10::Device const&) + 0x771 (0x5625037cfe31 in build/bin/test_api)
Nov 19 00:33:35 frame #11: FunctionalTest_isfinite_Test::TestBody() + 0x2f5 (0x5625037b2195 in build/bin/test_api)
Nov 19 00:33:35 frame #12: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x2bb (0x5625040aafdb in build/bin/test_api)
Nov 19 00:33:35 frame #13: testing::Test::Run() + 0x25b (0x562504059acb in build/bin/test_api)
Nov 19 00:33:35 frame #14: testing::TestInfo::Run() + 0x543 (0x56250405c413 in build/bin/test_api)
Nov 19 00:33:35 frame #15: testing::TestCase::Run() + 0x5fb (0x56250405e33b in build/bin/test_api)
Nov 19 00:33:35 frame #16: testing::internal::UnitTestImpl::RunAllTests() + 0x13d5 (0x562504080745 in build/bin/test_api)
Nov 19 00:33:35 frame #17: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x2b2 (0x5625040b0712 in build/bin/test_api)
Nov 19 00:33:35 frame #18: testing::UnitTest::Run() + 0x29a (0x56250407f1ea in build/bin/test_api)
Nov 19 00:33:35 frame #19: main + 0x458 (0x5625032bc328 in build/bin/test_api)
Nov 19 00:33:35 frame #20: __libc_start_main + 0xf0 (0x7fabdc9ea830 in /lib/x86_64-linux-gnu/libc.so.6)
Nov 19 00:33:35 frame #21: _start + 0x29 (0x5625032bbcd9 in build/bin/test_api)
Nov 19 00:33:35 " thrown in the test body.
Nov 19 00:33:35 [  FAILED  ] FunctionalTest.isfinite (12 ms)
```
We might need to split this test into `TEST_F(FunctionalTest, isfinite)` and `TEST_F(FunctionalTest, isfinite_CUDA)`, and the latter would contain only the CUDA tests.

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -1960,3 +1960,63 @@ TEST_F(FunctionalTest, Dropout3d) {
+    test_isfinite<torch::kInt8, char>(device);
+    test_isfinite<torch::kInt16, short>(device);
+    test_isfinite<torch::kInt32, int>(device);
+    test_isfinite<torch::kInt64, long>(device);
```
I suspect that `long` represents types of different bit-length on different platforms, and we might want to use the following list instead:
```
torch::kByte, uint8_t
torch::kChar, int8_t
torch::kShort, int16_t
torch::kInt, int
torch::kLong, int64_t
torch::kFloat, float
torch::kDouble, double
```

yf225 commented:
@pbelevich This PR seems to be breaking master with errors such as:
```
Nov 18 19:47:30 ../test/cpp/api/functional.cpp:2001:40: error: conversion from 'const long' to 'at::Scalar' is ambiguous
Nov 18 19:47:30     const auto x = torch::full({3, 3}, value, torch::TensorOptions().dtype(S).device(device));
Nov 18 19:47:30                                        ^~~~~
```
I will have to revert it for now, and I commented about the possible solutions in-line :D
